src/compiler/program.ts

```diff
@@ -2049,6 +2119,69 @@ namespace ts {
             return allFilesBelongToPath;
         }
 
+        function createProjectReferenceRedirects(rootOptions: CompilerOptions): Map<string> {
+            const result = createMap<string>();
+            const handledProjects = createMap<true>();
+
+            walkProjectReferenceGraph(host, rootOptions, createMapping);
+
+            function createMapping(resolvedFile: string, referencedProject: CompilerOptions, reference: ProjectReference) {
+                const rootDir = normalizePath(referencedProject.rootDir || getDirectoryPath(resolvedFile));
+                result.set(rootDir, referencedProject.outDir);
+                // If this project uses outFile, add the outFile .d.ts to our compilation
+                // Note: We don't enumerate the input files to try to find corresponding .d.ts
+                // files because we can't know from the outside whether they're modules or not
+                if (referencedProject.outFile) {
+                    const outFile = combinePaths(referencedProject.outDir, changeExtension(referencedProject.outFile, ".d.ts"));

```

Comment:
the outFile is not relative to the `outDir`. it is relative to the tsconfig.json file.. maybe i do not understand the rational here.

Reply:
it can also depend on declarationDir.